renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -514,6 +539,550 @@ struct D3D12PixelHistoryCallback : public D3D12ActionCallback
+  void PostRemisc(uint32_t eid, ActionFlags flags, ID3D12GraphicsCommandListX *cmd) {}
+  void PreCloseCommandList(ID3D12GraphicsCommandListX *cmd) {}
+  void AliasEvent(uint32_t primary, uint32_t alias) {}
+
+  void FetchOcclusionResults()
+  {
+    if(m_OcclusionQueries.size() == 0)
+      return;
+
+    D3D12_HEAP_PROPERTIES heapProps;
+    heapProps.Type = D3D12_HEAP_TYPE_READBACK;
+    heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
+    heapProps.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
+    heapProps.CreationNodeMask = 1;
+    heapProps.VisibleNodeMask = 1;
+
+    D3D12_RESOURCE_DESC bufDesc;
+    bufDesc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
+    bufDesc.Alignment = 0;
+    bufDesc.Width = sizeof(uint64_t) * m_OcclusionQueries.size();
+    bufDesc.Height = 1;
+    bufDesc.DepthOrArraySize = 1;
+    bufDesc.MipLevels = 1;
+    bufDesc.Format = DXGI_FORMAT_UNKNOWN;
+    bufDesc.SampleDesc.Count = 1;
+    bufDesc.SampleDesc.Quality = 0;
+    bufDesc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;
+    bufDesc.Flags = D3D12_RESOURCE_FLAG_NONE;
+
+    ID3D12Resource *readbackBuf = NULL;
+    HRESULT hr = m_pDevice->CreateCommittedResource(&heapProps, D3D12_HEAP_FLAG_NONE, &bufDesc,
+                                                    D3D12_RESOURCE_STATE_COPY_DEST, NULL,
+                                                    __uuidof(ID3D12Resource), (void **)&readbackBuf);
+    m_pDevice->CheckHRESULT(hr);
+    if(FAILED(hr))
+    {
+      RDCERR("Failed to create query readback buffer HRESULT: %s", ToStr(hr).c_str());
+      return;
+    }
+
+    ID3D12GraphicsCommandListX *list = m_pDevice->GetNewList();
+    if(!list)
+      return;
+
+    list->ResolveQueryData(m_OcclusionQueryHeap, D3D12_QUERY_TYPE_OCCLUSION, 0,
+                           (UINT)m_OcclusionQueries.size(), readbackBuf, 0);
+
+    list->Close();
+
+    m_pDevice->ExecuteLists();
+    m_pDevice->FlushLists();
+
+    D3D12_RANGE range;
+    range.Begin = 0;
+    range.End = (SIZE_T)bufDesc.Width;
+
+    uint64_t *data;
+    hr = readbackBuf->Map(0, &range, (void **)&data);
+    m_pDevice->CheckHRESULT(hr);
+    if(FAILED(hr))
+    {
+      RDCERR("Failed to map query heap data HRESULT: %s", ToStr(hr).c_str());
+      SAFE_RELEASE(readbackBuf);
+      return;
+    }
+
+    m_OcclusionResults.resize(m_OcclusionQueries.size());
+    for(size_t i = 0; i < m_OcclusionResults.size(); ++i)
+      m_OcclusionResults[i] = data[i];
+
+    readbackBuf->Unmap(0, NULL);
+    SAFE_RELEASE(readbackBuf);
+  }
+
+  uint64_t GetOcclusionResult(uint32_t eventId)
+  {
+    auto it = m_OcclusionQueries.find(eventId);
+    if(it == m_OcclusionQueries.end())
+      return 0;
+    RDCASSERT(it->second < m_OcclusionResults.size());
+    return m_OcclusionResults[it->second];
+  }
+
+private:
+  ID3D12PipelineState *GetPixelOcclusionPipeline(uint32_t eid, D3D12RenderState &state,
+                                                 uint32_t outputIndex)
```

Comment:
I'm not sure without more test coverage what would be needed here - either a separate shader for the target RT index (which is what the Vulkan impl does) or configuring the PSO to mask out writes for RTs other than the selected one (which would be preferable since it doesn't need unique shaders). The existing tests, which are mirrored from the Vulkan ones, don't test multiple RT pixel history, yet.

Reply:
All that to say, I'll add a TODO. :)